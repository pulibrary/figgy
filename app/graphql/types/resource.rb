# frozen_string_literal: true
module Types::Resource
  include Types::BaseInterface
  description "A resource in the system."
  orphan_types Types::ScannedResourceType, Types::FileSetType, Types::ScannedMapType

  field :id, String, null: true
  field :label, String, null: true
  field :viewing_hint, String, null: true
  field :url, String, null: true
  field :members, [Types::Resource], null: true
  field :source_metadata_identifier, String, null: true
  field :thumbnail, Types::Thumbnail, null: true

  definition_methods do
    def resolve_type(object, _context)
      "Types::#{object.class}Type".constantize
    end
  end

  def members
    @members ||= Wayfinder.for(object).members
  end

  def manifest_url
    helper.polymorphic_url([:manifest, object])
  end

  def url
    @url ||= helper.show_url(object)
  end

  # We need to centralize logic for navigating a MVW's members to find a
  # thumbnail file set. This is a hack to use the helper's logic for doing that.
  # refactor ticketed as https://github.com/pulibrary/figgy/issues/1708
  def helper
    @helper ||= ManifestBuilder::ManifestHelper.new.tap do |helper|
      helper.singleton_class.include(ThumbnailHelper)
      helper.define_singleton_method(:image_tag) do |url, _opts|
        url
      end
      helper.define_singleton_method(:image_path) do |url|
        url
      end
    end
  end

  def thumbnail
    return if object.try(:thumbnail_id).blank? || thumbnail_resource.blank?
    {
      id: thumbnail_resource.id.to_s,
      thumbnail_url: helper.figgy_thumbnail_path(thumbnail_resource),
      iiif_service_url: helper.figgy_thumbnail_path(thumbnail_resource).gsub("/full/!200,150/0/default.jpg", "")
    }
  end

  def thumbnail_resource
    @thumbnail_resource ||= query_service.find_by(id: object.try(:thumbnail_id).first)
  rescue Valkyrie::Persistence::ObjectNotFoundError
    nil
  end

  def query_service
    Valkyrie::MetadataAdapter.find(:indexing_persister).query_service
  end
end
