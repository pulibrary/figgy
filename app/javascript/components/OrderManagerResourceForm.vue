<template>
  <div>
    <heading level="h2">
      Set Properties <small>
        for this <span v-if="isMultiVolume">
          multi-volume
        </span> resource
      </small>
    </heading>
    <span class="lux-file_count">
      <text-style variation="emphasis">
        Total files: {{ memberCount }}
      </text-style>
    </span>
    <span
      v-if="resource.bibId"
      class="lux-bibid"
    >
      | BibId: {{ resource.bibId }}
    </span>
    <form
      id="app"
      novalidate="true"
    >
      <input-radio
        id="viewDir"
        vertical
        group-label="Viewing Direction"
        :options="viewDirs"
        :value="viewDirs.value"
        @change="updateViewDir($event)"
      />
      <input-radio
        v-if="!isMultiVolume"
        id="viewHint"
        vertical
        group-label="Viewing Hint"
        :options="viewHints"
        :value="viewHints.value"
        @change="updateViewHint($event)"
      />
    </form>
  </div>
</template>

<script>
/**
 * This is the Resource Form for the Order Manager in Figgy
 */
import { mapState } from 'vuex'
export default {
  name: 'ResourceForm',
  status: 'ready',
  release: '1.0.0',
  type: 'Pattern',
  metaInfo: {
    title: 'Resource Form',
    htmlAttrs: {
      lang: 'en'
    }
  },
  props: {
    /**
     * The html element name used for the component.
     */
    type: {
      type: String,
      default: 'div'
    },
    count: {
      type: Number,
      default: 0
    }
  },
  computed: {
    memberCount: function () {
      return this.$store.getters.getMemberCount
    },
    isMultiVolume () {
      return this.$store.getters.isMultiVolume
    },
    ...mapState({
      resource: state => state.ordermanager.resource
    }),
    viewHints: function () {
      return [
        {
          name: 'viewHint',
          value: 'individuals',
          id: 'individuals',
          checked: this.resource.viewingHint === 'individuals'
        },
        { name: 'viewHint', value: 'paged', id: 'paged', checked: this.resource.viewingHint === 'paged' },
        {
          name: 'viewHint',
          value: 'continuous',
          id: 'continuous',
          checked: this.resource.viewingHint === 'continuous'
        }
      ]
    },
    viewDirs: function () {
      return [
        {
          name: 'viewDir',
          value: 'LEFTTORIGHT',
          id: 'left-to-right',
          label: 'left-to-right',
          checked: this.resource.viewingDirection === 'LEFTTORIGHT'
        },
        {
          name: 'viewDir',
          value: 'RIGHTTOLEFT',
          id: 'right-to-left',
          label: 'right-to-left',
          checked: this.resource.viewingDirection === 'RIGHTTOLEFT'
        },
        {
          name: 'viewDir',
          value: 'TOPTOBOTTOM',
          id: 'top-to-bottom',
          label: 'top-to-bottom',
          checked: this.resource.viewingDirection === 'TOPTOBOTTOM'
        },
        {
          name: 'viewDir',
          value: 'BOTTOMTOTOP',
          id: 'bottom-to-top',
          label: 'bottom-to-top',
          checked: this.resource.viewingDirection === 'BOTTOMTOTOP'
        }
      ]
    }
  },
  methods: {
    isIndividuals: function () {
      return this.resource.viewingHint === 'individuals'
    },
    isPaged: function () {
      return this.resource.viewingHint === 'paged'
    },
    isContinuous: function () {
      return this.resource.viewingHint === 'continuous'
    },
    updateViewDir (value) {
      this.$store.dispatch('updateViewDir', value)
    },
    updateViewHint (value) {
      this.$store.dispatch('updateViewHint', value)
    }
  }
}
</script>

<style lang="scss" scoped>
small {
  font-size: 1rem;
  font-weight: 400;
}
</style>
