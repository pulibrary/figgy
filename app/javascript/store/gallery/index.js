export const galleryState = {
  items: [],
  selected: [],
  cut: [],
  changeList: [],
  ogItems: [],
}

export const galleryMutations = {
  CUT(state, itemArray) {
    state.cut = [...itemArray]
  },
  PASTE(state, itemArray) {
    state.items = [...itemArray]
  },
  SELECT(state, itemArray) {
    state.selected = [...itemArray]
  },
  SET_GALLERY(state, items) {
    state.items = items
    state.ogItems = items
  },
  SORT_ITEMS(state, value) {
    state.items = [...value]
  },
  UPDATE_CHANGES(state, changeList) {
    state.changeList = [...changeList]
  },
  UPDATE_ITEMS(state, items) {
    state.items = [...items]
  },
}

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}
