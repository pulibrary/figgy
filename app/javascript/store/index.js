import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import actions from './vuex/actions'
import {modules} from 'lux-design-system'
Vue.use(Vuex)

const store = new Vuex.Store({
  actions,
  modules: {
    ordermanager: modules.resourceModule,
    gallery: modules.galleryModule,
  }
})

export default store
