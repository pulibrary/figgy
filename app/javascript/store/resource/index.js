export const resourceState = {
  resource: {
    id: "",
    resourceClassName: "",
    bibId: "",
    label: "Resource not available.",
    thumbnail: "",
    startCanvas: "",
    isMultiVolume: false,
    viewingHint: null,
    viewingDirection: null,
    members: [],
    loadState: "NOT_LOADED",
    saveState: "NOT_SAVED",
    ogState: {},
  },
}

export const resourceMutations = {
  APPLY_STATE(state) {
    state.gallery.ogItems = state.gallery.items
    state.gallery.changeList = []
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
    state.resource.saveState = "NOT_SAVED"
  },
  CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
    state.resource.loadState = loadState
  },
  SAVED_STATE(state, saveStatus) {
    state.resource.saveState = saveStatus
  },
  SET_RESOURCE(state, resource) {
    state.resource.id = resource.id
    state.resource.resourceClassName = resource.__typename
    state.resource.label = resource.label
    state.resource.startCanvas = resource.startPage
    state.resource.viewingHint = resource.viewingHint
    state.resource.viewingDirection = resource.viewingDirection
    state.resource.thumbnail = resource.thumbnail != null ? resource.thumbnail.id : null
    state.resource.members = resource.members

    const items = resource.members.map(member => ({
      id: member.id,
      viewingHint: member.viewingHint != null ? member.viewingHint : "single",
      caption: member.label, // member.__typename + " : " + member.id,
      service:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl
          : Global.figgy.resource.defaultThumbnail,
      mediaUrl:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
          : Global.figgy.resource.defaultThumbnail,
    }))
    state.gallery.items = items
    state.gallery.ogItems = items
    state.resource.loadState = "LOADED"
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail != null ? resource.thumbnail.id : null,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
  },
  UPDATE_STARTCANVAS(state, startCanvas) {
    state.resource.startCanvas = startCanvas
  },
  UPDATE_THUMBNAIL(state, thumbnail) {
    state.resource.thumbnail = thumbnail
  },
  UPDATE_VIEWDIR(state, viewDir) {
    state.resource.viewingDirection = viewDir
  },
  UPDATE_VIEWHINT(state, viewHint) {
    state.resource.viewingHint = viewHint
  },
}

export const resourceGetters = {
  getMemberCount: state => {
    return state.resource.members.length
  },
  isMultiVolume: state => {
    const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
    return volumes.length > 0 ? true : false
  },
  orderChanged: state => {
    let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
    let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
    return ogOrder !== imgOrder
  },
  stateChanged: (state, getters) => {
    var propsChanged = []
    propsChanged.push(state.resource.ogState.thumbnail !== state.resource.thumbnail)
    propsChanged.push(state.resource.ogState.startCanvas !== state.resource.startCanvas)
    propsChanged.push(state.resource.ogState.viewingHint !== state.resource.viewingHint)
    propsChanged.push(state.resource.ogState.viewingDirection !== state.resource.viewingDirection)
    propsChanged.push(state.gallery.changeList.length > 0)
    propsChanged.push(getters.orderChanged)
    if (propsChanged.indexOf(true) > -1) {
      return true
    } else {
      return false
    }
  },
}
