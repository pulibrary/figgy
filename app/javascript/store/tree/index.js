export const treeState = {
  selected: null,
}

export const treeMutations = {
  SELECT(state, selectedId) {
    state.selected = selectedId
  },
}
