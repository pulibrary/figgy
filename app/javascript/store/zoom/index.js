export const zoomState = {
  zoomed: null,
}

export const zoomMutations = {
  ZOOM(state, obj) {
    state.zoomed = obj
  },
  RESET_ZOOM(state) {
    state.zoomed = null
  },
}

export const zoomGetters = {
  getIsZoomed: state => {
    return state.zoomed
  },
}
