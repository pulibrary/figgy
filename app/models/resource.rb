# frozen_string_literal: true
class Resource < Valkyrie::Resource
  def self.human_readable_type
    default = @_human_readable_type || name.demodulize.titleize
    I18n.translate("models.#{new.model_name.i18n_key}", default: default)
  end

  def self.can_have_manifests?
    false
  end

  def human_readable_type
    self.class.human_readable_type
  end

  def self.model_name
    ::ActiveModel::Name.new(self)
  end

  # Determines if this is an image resource
  # @return [TrueClass, FalseClass]
  def image_resource?
    false
  end

  # Determines if this is a geospatial resource
  # @return [TrueClass, FalseClass]
  def geo_resource?
    false
  end

  # Determines if this is a media resource
  # @return [TrueClass, FalseClass]
  def media_resource?
    false
  end

  # Virtual property used for stashing pre-loaded objects. Populated by a query.
  attr_accessor :loaded
  def loaded
    @loaded ||= {}
  end
end
