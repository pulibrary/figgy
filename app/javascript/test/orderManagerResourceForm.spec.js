import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import OrderManagerResourceForm from "../components/OrderManagerResourceForm.vue"
import { resourceMutations, resourceGetters } from "../store/resource"
import { modules } from 'lux-design-system'

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.use(Vuex)

let wrapper
let getters
let options
let store
let actions
let items = [
  { id: "a", title: "A", viewingHint: "single", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]

describe("OrderManagerResourceForm.vue", () => {
  beforeEach(() => {
    actions = {
      updateViewHint: jest.fn(),
      updateViewDir: jest.fn()
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const getters = {
      getMemberCount: () => 3
    }

    const resource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "4609321",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: "single",
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: resource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        viewingDirection: function () {
          return 'bottom-to-top'
        },
        viewingHint: function () {
          return 'continuous'
        }
      }
    }

    wrapper = mount(OrderManagerResourceForm, {
      options,
      localVue,
      store,
      stubs: ["heading","input-radio","text-style"],
    })
  })

  it('allows the selection of a viewing direction', () => {
    wrapper.find('#viewDir').trigger('change')
    wrapper.vm.updateViewDir('LEFTTORIGHT')
    expect(actions.updateViewDir).toHaveBeenCalled()
  })

  it('allows the selection of a viewing hint', () => {
    wrapper.vm.updateViewHint('individuals')
    expect(actions.updateViewHint).toHaveBeenCalled()
  })

  it('displays the correct image (aka, fileset) count', () => {
    const expanded = wrapper.find('.lux-file_count')
    const fileCount = expanded.text()
    expect(fileCount).toEqual('Total files: 3')
  })

  it('displays the bibid, if it has one', () => {
    const expanded = wrapper.find('.lux-bibid')
    const bibid = expanded.text()
    expect(bibid).toEqual('| BibId: 4609321')
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
