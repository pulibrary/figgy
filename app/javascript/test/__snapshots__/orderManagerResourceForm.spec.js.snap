// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManagerResourceForm.vue has the expected html structure 1`] = `
<div>
  <heading-stub
    level="h2"
  >
    
    Set Properties 
    <small>
      
      for this 
      <!---->
       resource
    
    </small>
  </heading-stub>
   
  <span
    class="lux-file_count"
  >
    <text-style-stub
      variation="emphasis"
    >
      
      Total files: 3
    
    </text-style-stub>
  </span>
   
  <span
    class="lux-bibid"
  >
    
    | BibId: 4609321
  
  </span>
   
  <form
    id="app"
    novalidate="novalidate"
  >
    <input-radio-stub
      group-label="Viewing Direction"
      id="viewDir"
      options="[object Object],[object Object],[object Object],[object Object]"
      vertical=""
    />
     
    <input-radio-stub
      group-label="Viewing Hint"
      id="viewHint"
      options="[object Object],[object Object],[object Object]"
      vertical=""
    />
  </form>
</div>
`;
